src/main/scala/io/dronekit/cloud/ShellCommandFlow.scala

```diff
@@ -0,0 +1,215 @@
+    val in: Inlet[ByteString] = Inlet("CommandInput")
+    val out: Outlet[String] = Outlet("CommandOutput")
+    override val shape: FlowShape[ByteString, String] = FlowShape(in, out)
+
+    sealed trait Output
+    case object StandardOutput extends Output
+    case class ErrorOutput(out: String) extends Output
+    case object StandardOutputFinished extends Output
+
+
+    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+      new GraphStageLogic(shape) {
+        val log: Logger = Logger(LoggerFactory.getLogger(getClass))
+        var upstreamFinished = new AtomicBoolean(false)
+        var inputClosed = new AtomicBoolean(false)
+        var outputClosed = new AtomicBoolean(false)
+        var errorAbort = new AtomicBoolean(false)
+        val process = Process(command)
+        val inputQueue = new LinkedBlockingDeque[ByteString]()
+        val outputQueue = new LinkedBlockingDeque[String]()
+        val errorQueue = new LinkedBlockingDeque[String]()
+        val callback = getAsyncCallback(outputAsyncInput)
+
+        override def preStart(): Unit = {
+          val io = new ProcessIO(processInput, processOutput, processErrorOutput)
+          // TODO: create a materialized value with the exit code
+          process.run(io)
+        }
+
+        private def pushAndPull(): Unit = {
+          if (outputQueue.isEmpty && !hasBeenPulled(in) && !isClosed(in)) {
+            log.debug("pulling")
+            pull(in)
+          }
+          while (!outputQueue.isEmpty && !isClosed(out) && isAvailable(out)) {
+            val output = outputQueue.takeLast()
+            log.debug(s"pushing $output")
+            push(out, output)
+          }
+        }
+
+        private def pushAndComplete(): Unit = {
+          if (outputQueue.isEmpty) {
+            log.info("Completing stage")
+            completeStage()
+          }
+          else {
+            if (isAvailable(out)) {
+              val finalPush = outputQueue.descendingIterator().mkString("\n")
+              log.info(s"Pushing final value and completing")
+              push(out, finalPush)
+              completeStage()
+            }
+            else {
+              // Wait for onPull
+              log.info("waiting for onPull to complete stage")
+            }
+          }
+        }
+
+        private def outputAsyncInput(data: Output): Unit = data match {
+          case StandardOutput =>
+            pushAndPull()
+          case ErrorOutput(msg) =>
+            log.error(msg)
+            errorQueue.putFirst(msg)
+          case StandardOutputFinished =>
+            log.debug("StandardOutputFinished")
+            outputClosed.set(true)
+            pushAndComplete()
+
+        }
+
+        setHandler(in, new InHandler{
+          override def onPush: Unit = {
+            val elem = grab(in)
+            if (!errorQueue.isEmpty) {
+              // What is the right thing to do if there is stderr output?
```

Comment:
This is out of scope for this PR but eventually the user should be able to consume stderr as a different stream and decide what to do with it.